File: backend/routes/notifications.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Optional
from supabase_client import supabase_admin

router = APIRouter(prefix="/api", tags=["notifications"])

class CreateNotificationRequest(BaseModel):
    user_id: str
    title: str
    message: str
    type: Optional[str] = "system"

class CreateReminderRequest(BaseModel):
    user_id: str
    title: str
    description: Optional[str] = ""
    reminder_type: str
    time: str
    days: list = []

@router.get("/user/{user_id}/notifications")
async def get_notifications(user_id: str):
    try:
        response = supabase_admin.table("notifications").select("*").eq("user_id", user_id).order("timestamp", desc=True).limit(50).execute()

        return response.data
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to get notifications: {str(e)}")

@router.post("/notifications")
async def create_notification(req: CreateNotificationRequest):
    try:
        notification_data = {
            "user_id": req.user_id,
            "title": req.title,
            "message": req.message,
            "type": req.type
        }

        response = supabase_admin.table("notifications").insert(notification_data).execute()

        if not response.data:
            raise HTTPException(status_code=500, detail="Failed to create notification")

        return {"message": "Notification created successfully", "notification": response.data[0]}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to create notification: {str(e)}")

@router.patch("/notifications/{notification_id}/read")
async def mark_notification_read(notification_id: str):
    try:
        response = supabase_admin.table("notifications").update({"read": True}).eq("id", notification_id).execute()

        if not response.data:
            raise HTTPException(status_code=404, detail="Notification not found")

        return {"message": "Notification marked as read"}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to mark notification as read: {str(e)}")

@router.get("/user/{user_id}/reminders")
async def get_reminders(user_id: str):
    try:
        response = supabase_admin.table("reminders").select("*").eq("user_id", user_id).execute()

        return response.data
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to get reminders: {str(e)}")

@router.post("/reminders")
async def create_reminder(req: CreateReminderRequest):
    try:
        reminder_data = {
            "user_id": req.user_id,
            "title": req.title,
            "description": req.description,
            "reminder_type": req.reminder_type,
            "time": req.time,
            "days": req.days
        }

        response = supabase_admin.table("reminders").insert(reminder_data).execute()

        if not response.data:
            raise HTTPException(status_code=500, detail="Failed to create reminder")

        return {"message": "Reminder created successfully", "reminder": response.data[0]}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to create reminder: {str(e)}")

@router.patch("/reminders/{reminder_id}")
async def update_reminder(reminder_id: str, data: dict):
    try:
        response = supabase_admin.table("reminders").update(data).eq("id", reminder_id).execute()

        if not response.data:
            raise HTTPException(status_code=404, detail="Reminder not found")

        return {"message": "Reminder updated successfully", "reminder": response.data[0]}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to update reminder: {str(e)}")

@router.delete("/reminders/{reminder_id}")
async def delete_reminder(reminder_id: str):
    try:
        response = supabase_admin.table("reminders").delete().eq("id", reminder_id).execute()

        if not response.data:
            raise HTTPException(status_code=404, detail="Reminder not found")

        return {"message": "Reminder deleted successfully"}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to delete reminder: {str(e)}")

File: backend/routes/notifications.js
import express from 'express';
import nodemailer from 'nodemailer';
import supabase from '../lib/supabase.js';
import authenticateUser from '../middleware/auth.js';

const router = express.Router();

const transporter = nodemailer.createTransport({
  host: process.env.EMAIL_HOST || 'smtp.gmail.com',
  port: parseInt(process.env.EMAIL_PORT || '587'),
  secure: false,
  auth: {
    user: process.env.EMAIL_USER,
    pass: process.env.EMAIL_PASSWORD
  }
});

router.get('/', authenticateUser, async (req, res) => {
  try {
    const userId = req.user.id;
    const { limit = 20, unread_only = false } = req.query;

    let query = supabase
      .from('notifications')
      .select('*')
      .eq('user_id', userId)
      .order('created_at', { ascending: false })
      .limit(limit);

    if (unread_only === 'true') {
      query = query.eq('read', false);
    }

    const { data: notifications, error } = await query;

    if (error) throw error;

    res.json({ notifications });
  } catch (error) {
    console.error('Fetch notifications error:', error);
    res.status(500).json({ error: 'Failed to fetch notifications' });
  }
});

router.post('/', authenticateUser, async (req, res) => {
  try {
    const { user_id, title, message, type } = req.body;

    if (!user_id || !title || !message || !type) {
      return res.status(400).json({ error: 'Missing required fields' });
    }

    const { data: notification, error } = await supabase
      .from('notifications')
      .insert({
        user_id,
        title,
        message,
        type
      })
      .select()
      .single();

    if (error) throw error;

    res.status(201).json({
      message: 'Notification created successfully',
      notification
    });
  } catch (error) {
    console.error('Create notification error:', error);
    res.status(500).json({ error: 'Failed to create notification' });
  }
});

router.put('/:notificationId/read', authenticateUser, async (req, res) => {
  try {
    const { notificationId } = req.params;
    const userId = req.user.id;

    const { data: notification, error } = await supabase
      .from('notifications')
      .update({ read: true })
      .eq('id', notificationId)
      .eq('user_id', userId)
      .select()
      .single();

    if (error) throw error;

    res.json({
      message: 'Notification marked as read',
      notification
    });
  } catch (error) {
    console.error('Mark read error:', error);
    res.status(500).json({ error: 'Failed to mark notification as read' });
  }
});

router.put('/mark-all-read', authenticateUser, async (req, res) => {
  try {
    const userId = req.user.id;

    const { error } = await supabase
      .from('notifications')
      .update({ read: true })
      .eq('user_id', userId)
      .eq('read', false);

    if (error) throw error;

    res.json({ message: 'All notifications marked as read' });
  } catch (error) {
    console.error('Mark all read error:', error);
    res.status(500).json({ error: 'Failed to mark all as read' });
  }
});

router.post('/send-email', async (req, res) => {
  try {
    const { to, subject, html, text } = req.body;

    if (!to || !subject || (!html && !text)) {
      return res.status(400).json({ error: 'Missing required fields' });
    }

    const mailOptions = {
      from: process.env.EMAIL_FROM || 'noreply@levelup.app',
      to,
      subject,
      text,
      html
    };

    await transporter.sendMail(mailOptions);

    res.json({ message: 'Email sent successfully' });
  } catch (error) {
    console.error('Send email error:', error);
    res.status(500).json({ error: 'Failed to send email' });
  }
});

export default router;
